File: 程序员面试金典/面试题 04.06. 后继者/cpp/main3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    TreeNode *inorderSuccessor(TreeNode *root, TreeNode *p)
    {
        if (p->right)
        {
            p = p->right;
            while (p->left) p = p->left;
            return p;
        }
        TreeNode* ret = nullptr;
        while (root != p)
        {
            if (root->val < p->val)
            {
                root = root->right;
            }
            else
            {
                ret = root;
                root = root->left;
            }
        }
        return ret;
    }
};

File: 程序员面试金典/面试题 04.06. 后继者/Java/main3.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode inorderSuccessor(TreeNode root, TreeNode p) {
        if (p.right != null) {
            p = p.right;
            while (p.left != null) {
                p = p.left;
            }
            return p;
        }
        TreeNode ret = null;
        while (root != p) {
            if (root.val < p.val) {
                root = root.right;
            }
            else {
                ret = root;
                root = root.left;
            }
        }
        return ret;
    }
}